internal/pkg/deploy/deploy.go

```diff
@@ -18,22 +24,130 @@ type CreateEnvironmentInput struct {
 	ToolsAccountPrincipalARN string // The Principal ARN of the tools account.
 }
 
+const (
+	GithubProviderName = "GitHub"
+	GithubSecretIdKeyName = "githubOAuthSecretId"
+)
+
 // CreatePipelineInput represents the fields required to deploy a pipeline.
 type CreatePipelineInput struct {
 	// Name of the project this pipeline belongs to
 	ProjectName string
 	// Name of the pipeline
 	Name   string
+	// The source code provider for this pipeline
 	Source *Source
+	// The stages of the pipeline. The order of stages in this list
+	// will be the order we deploy to
 	Stages []PipelineStage
+	// A list of artifact buckets and corresponding KMS keys that will
+	// be used in this pipeline.
+	ArtifactBuckets []ArtifactBucket
+}
+
+// ArtifactBucket represents an S3 bucket used by the CodePipeline to store
+// intermediate artifacts produced by the pipeline.
+type ArtifactBucket struct {
+	// The ARN of the S3 bucket.
+	BucketArn string
+	// The ARN of the KMS key used to en/decrypt artifacts stored in this bucket.
+	KeyArn string
+}
+
+// Region parses out the region from the ARN of the KMS key associated with
+// the artifact bucket. 
+func (a *ArtifactBucket) Region() (string, error) {
+	// We assume the bucket and the key are in the same AWS region.
+	parsedArn, err := arn.Parse(a.KeyArn)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse region out of key ARN: %s, error: %w",
+		a.BucketArn, err)
+	}
+	return parsedArn.Region, nil
+}
+
+// BucketName parses out the name of the bucket from its ARN.
+func (a *ArtifactBucket) BucketName() (string, error) {
+	parsedArn, err := arn.Parse(a.BucketArn)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse the name of the bucket out of bucket ARN: %s, error: %w",
+		a.BucketArn, err)
+	}
+	return parsedArn.Resource, nil
 }
 
 // Source defines the source of the artifacts to be built and deployed.
 type Source struct {
+	// The name of the source code provider. For example, "GitHub"
 	ProviderName string
+	
+	// Contains provider-specific configurations, such as:
+	// "repository": "aws/amazon-ecs-cli-v2"
+	// "githubOAuthSecretId": "heyyo"
 	Properties   map[string]interface{}
 }
 
+// GitHubOAuthSecretID returns the ID of the secret in the Secrets manager,
+// which stores the GitHub OAuth token if the provider is "GitHub". Otherwise,
+// it returns an error.
+func (s *Source) GitHubOAuthSecretID() (string, error) {
+	secretID, exists := s.Properties[GithubSecretIdKeyName]
+	if !exists {
+		return "", errors.New("the GitHub token secretID is not configured")
+	}
+
+	id, ok := secretID.(string)
+	if !ok {
+		return "", fmt.Errorf("unable to locate the GitHub token secretID from %v", secretID)
+	}
+
+	if s.ProviderName != GithubProviderName {
+		return "", fmt.Errorf("failed attempt to retrieve GitHub token from a non-GitHub provider")
+	}
+
+	return id, nil
+}
+
+func (s *Source) parseOwnerAndRepo() (string, string, error) {
```

Comment:
I got this feedback in a past PR I agree that it's more readable: https://github.com/aws/amazon-ecs-cli-v2/pull/279#discussion_r342383190

What do you think of returning:
```go
type ownerAndRepo struct {
   owner string
   repo string
}
```